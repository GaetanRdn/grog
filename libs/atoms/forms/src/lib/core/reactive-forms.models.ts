export type OnChangeFn<ValueType> = (value: ValueType | null) => void;
export type OnTouchedFn = () => void;
