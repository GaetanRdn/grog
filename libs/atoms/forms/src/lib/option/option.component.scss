:host {
  flex: 1;
  padding: calc(var(--default-padding) * 2) calc(var(--default-padding) * 4);
}

@mixin gro-option($config) {
  $typography: map-get(map-get($config, typography), option);
  $colors: map-get($config, colors);
  $colorKeys: ['primary', 'accent', 'error'];

  gro-option {
    font-family: map-get($typography, font-family);
    font-size: map-get($typography, font-size);
    line-height: map-get($typography, line-height);
    font-weight: map-get($typography, font-weight);
    color: map-get(map-get($colors, default), color);
    flex: 1 1 100%;

    &:not(.gro-disabled) {
      cursor: pointer;

      &:hover {
        background: transparentize(map-get(map-get($colors, default), color), 0.9);
      }

      &.gro-selected {
        color: map-get(map-get($colors, primary), color);
        border-left: 0.4rem inset map-get(map-get($colors, primary), color);
      }
    }

    &.gro-disabled {
      color: map-get(map-get($colors, disabled), color);
      background: map-get(map-get($colors, disabled), background);
      border: 0.1rem inset map-get(map-get($colors, disabled), border);

      &.gro-selected {
        border-left: 0.4rem inset map-get(map-get($colors, disabled), border);
      }
    }
  }
}
