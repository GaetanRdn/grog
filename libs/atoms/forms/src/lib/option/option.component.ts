import {
  ChangeDetectionStrategy,
  Component,
  ElementRef,
  EventEmitter,
  HostListener,
  Input,
  NgModule,
  OnInit,
  Output,
} from '@angular/core';
import { CoerceBoolean } from '@grorg/decorators';
import { BooleanInput } from '@grorg/types';

@Component({
  selector: 'gro-option',
  host: {
    '[class.gro-selected]': 'selected',
    '[class.gro-disabled]': 'disabled || null',
  },
  template: ` <ng-content></ng-content>`,
  styleUrls: ['./option.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class OptionComponent<ValueType> implements OnInit {
  static ngAcceptInputType_disabled: BooleanInput;

  @Input()
  public value!: ValueType;

  @Input()
  @CoerceBoolean()
  public disabled = false;

  @Output()
  public readonly selectedChange: EventEmitter<OptionComponent<ValueType>> = new EventEmitter<
    OptionComponent<ValueType>
  >();

  public selected = false;

  constructor(private readonly _elementRef: ElementRef<HTMLElement>) {}

  public ngOnInit(): void {
    if (!this.value) {
      this.value = this._elementRef.nativeElement.textContent as unknown as ValueType;
    }
  }

  @HostListener('click', ['$event'])
  public select(event: MouseEvent): void {
    if (!this.disabled && !this.selected) {
      this.selected = true;
      this.selectedChange.emit(this);
    }

    event.stopPropagation();
    event.stopImmediatePropagation();
  }
}

@NgModule({
  declarations: [OptionComponent],
  exports: [OptionComponent],
})
export class OptionModule {}
