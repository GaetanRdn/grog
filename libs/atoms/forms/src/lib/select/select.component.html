<span class="gro-select-wrapper">
<input [disabled]="disabled" readonly value="{{value}}" />
  <span class="gro-select-icon">keyboard_arrow_down</span>
</span>
<div *ngIf="isOpen" class="gro-options">
  <span (click)="select($event, option)" *ngFor="let option of options"
        [class.gro-selected]="value === option" class="gro-option">{{ option }}</span>
</div>
