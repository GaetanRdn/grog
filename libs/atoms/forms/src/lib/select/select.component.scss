:host {
  display: inline-flex;
  width: fit-content;
  min-width: 20rem;
  user-select: none;

  &:not(.gro-disabled) {
    cursor: pointer;
  }

  &.gro-opened {
    .gro-select-wrapper {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }

  .gro-select-wrapper {
    z-index: 2;
    border-radius: var(--default-radius);
    display: inline-flex;
    width: 100%;
    justify-content: space-between;
    align-items: center;
    padding: calc(var(--default-padding) * 3) calc(var(--default-padding) * 4);
  }

  .gro-options-wrapper {
    z-index: 1;
    width: 100%;
    left: 0;
    top: 100%;
    bottom: 100%;
    position: absolute;
    display: flex;
    flex-flow: column;
    border-bottom-left-radius: var(--default-radius);
    border-bottom-right-radius: var(--default-radius);
    overflow: hidden;
  }
}

@mixin gro-select($config) {
  $typography: map-get(map-get($config, typography), input);
  $colors: map-get($config, colors);

  gro-select {
    font-family: map-get($typography, font-family);
    font-size: map-get($typography, font-size);
    line-height: map-get($typography, line-height);
    font-weight: map-get($typography, font-weight);
    color: map-get(map-get($colors, default), color);

    .gro-options-wrapper, .gro-select-wrapper {
      border-left: 0.1rem inset map-get(map-get($colors, default), border);
      border-right: 0.1rem inset map-get(map-get($colors, default), border);
      border-bottom: 0.1rem inset map-get(map-get($colors, default), border);
      background-color: map-get(map-get($colors, default), background);
    }

    .gro-select-wrapper {
      border-top: 0.1rem inset map-get(map-get($colors, default), border);
    }

    .gro-select-wrapper:hover {
      background: transparentize(map-get(map-get($colors, default), color), 0.9);
    }

    &.gro-disabled {
      color: map-get(map-get($colors, disabled), color);

      .gro-select-wrapper {
        border: 0.1rem inset map-get(map-get($colors, disabled), border);
        background-color: map-get(map-get($colors, disabled), background);
      }
    }
  }
}
