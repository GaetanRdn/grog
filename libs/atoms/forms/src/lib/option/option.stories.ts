import { Args, Meta, moduleMetadata, Story } from '@storybook/angular';
import { OptionComponent, OptionModule } from './option.component';

export default {
  component: OptionComponent,
  title: 'atoms/forms/option',
  decorators: [
    moduleMetadata({
      imports: [OptionModule],
    }),
  ],
  argTypes: { selectedChange: { action: 'clicked' } },
} as Meta<OptionComponent<string>>;

const template: Story = (args: Args) => ({
  props: {
    disabled: Boolean(args['disabled']),
  },
  template: `<gro-option (selectedChange)="selectedChange" [disabled]="disabled">Option</gro-option>`,
});

export const Default = template.bind({});

export const Disabled = template.bind({});
Disabled.args = {
  disabled: true,
};
