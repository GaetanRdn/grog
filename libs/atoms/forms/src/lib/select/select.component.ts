import {
  AfterContentInit,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ContentChildren,
  ElementRef,
  EventEmitter,
  HostBinding,
  HostListener,
  Input,
  NgModule,
  NgZone,
  Optional,
  Output,
  QueryList,
} from '@angular/core';
import { OptionComponent, OptionModule } from '../option/option.component';
import { MaterialIconModule } from '@grorg/atoms/icons';
import { CommonModule } from '@angular/common';
import { defer, merge, Observable } from 'rxjs';
import { startWith, switchMap, take, tap } from 'rxjs/operators';
import { BooleanInput, EqualsFn, Nullable, OnChangeFn, OnTouchedFn, TypedControlValueAccessor } from '@grorg/types';
import { NgControl } from '@angular/forms';
import { CoerceBoolean } from '@grorg/decorators';
import { slideDownUp } from '@grorg/atoms/animations';

// TODO handle multiple selection
// TODO revoir la bordure en selected
@Component({
  selector: 'gro-select',
  animations: [slideDownUp],
  template: ` <div class="gro-select-wrapper">
      <span>{{ viewValue }}</span>
      <gro-material-icon>expand_more</gro-material-icon>
    </div>
    <div class="gro-options-wrapper" [style.height.rem]="panelHeight" *ngIf="opened" @slideDownUp>
      <ng-content></ng-content>
    </div>`,
  styleUrls: ['./select.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SelectComponent<ValueType> implements AfterContentInit, TypedControlValueAccessor<ValueType> {
  static ngAcceptInputType_disabled: BooleanInput;

  @Input()
  public placeholder?: string;

  @Input()
  public value: Nullable<ValueType> = null;

  @Output()
  public readonly valueChange: EventEmitter<Nullable<ValueType>> = new EventEmitter<Nullable<ValueType>>();

  @ContentChildren(OptionComponent)
  public readonly options!: QueryList<OptionComponent<ValueType>>;

  @HostBinding('class.gro-opened')
  public opened = false;

  private readonly _selectionChange$: Observable<OptionComponent<ValueType>> = defer(() => {
    const options = this.options;

    if (options) {
      return options.changes.pipe(
        startWith(options),
        switchMap(() => merge(...options.map((option) => option.selectedChange)))
      );
    }

    return this._ngZone.onStable.pipe(
      take(1),
      switchMap(() => this._selectionChange$)
    );
  });

  constructor(
    private readonly _elementRef: ElementRef,
    private readonly _ngZone: NgZone,
    private readonly _cdr: ChangeDetectorRef,
    @Optional() private readonly _ngControl: NgControl
  ) {
    if (_ngControl) {
      _ngControl.valueAccessor = this;
    }
  }

  private _disabled = false;

  get disabled(): boolean {
    return this._disabled || Boolean(this._ngControl?.disabled);
  }

  @HostBinding('class.gro-disabled')
  @Input()
  @CoerceBoolean()
  set disabled(disabled: boolean) {
    this._disabled = disabled;
  }

  private _viewValue: Nullable<string> = null;

  get viewValue(): string {
    return this._viewValue ?? this.placeholder ?? '';
  }

  get panelHeight(): number {
    return this.options.toArray().length * 4;
  }

  @Input()
  public equals: EqualsFn<Nullable<ValueType>> = (opt1: Nullable<ValueType>, opt2: Nullable<ValueType>): boolean =>
    opt1 === opt2;

  @HostListener('click')
  public openPanel(): void {
    if (!this.disabled) {
      this.opened = true;
    }
  }

  @HostListener('document:click', ['$event.target'])
  public onClick(event: EventTarget): void {
    if (!this.disabled && !this._elementRef.nativeElement.contains(event)) {
      this.opened = false;
    }
  }

  public ngAfterContentInit(): void {
    this._selectionChange$
      .pipe(
        tap((option: OptionComponent<ValueType>) =>
          this.options
            .toArray()
            .filter((opt: OptionComponent<ValueType>) => !this.equals(opt.value, option.value))
            .forEach((opt: OptionComponent<ValueType>) => (opt.selected = false))
        ),
        tap(() => (this.opened = false)),
        tap((option: OptionComponent<ValueType>) => (this._viewValue = option.value as unknown as string)),
        tap((option: OptionComponent<ValueType>) => (this.value = option.value)),
        tap(() => this.valueChange.emit(this.value)),
        tap(() => this._onChange(this.value)),
        tap(() => this._onTouched())
      )
      .subscribe(() => this._cdr.markForCheck());

    if (this.value) {
      this._viewValue = this.value as unknown as string;
      const found: OptionComponent<ValueType> | undefined = this.options.find((option: OptionComponent<ValueType>) =>
        this.equals(option.value, this.value)
      );

      if (found) {
        Promise.resolve().then(() => (found.selected = true));
      }
    }
  }

  public registerOnChange(fn: OnChangeFn<ValueType>): void {
    this._onChange = fn;
  }

  public registerOnTouched(fn: OnTouchedFn): void {
    this._onTouched = fn;
  }

  public writeValue(value: Nullable<ValueType>): void {
    this.value = value;
    this._viewValue = value as unknown as string;

    Promise.resolve().then(() => {
      const found: OptionComponent<ValueType> | undefined = this.options.find((option: OptionComponent<ValueType>) =>
        this.equals(option.value, this.value)
      );

      if (found) {
        found.selected = true;
      }
    });
  }

  private _onChange: OnChangeFn<ValueType> = () => ({});

  private _onTouched: OnTouchedFn = () => ({});
}

@NgModule({
  declarations: [SelectComponent],
  exports: [SelectComponent, OptionModule],
  imports: [MaterialIconModule, CommonModule],
})
export class SelectModule {}
