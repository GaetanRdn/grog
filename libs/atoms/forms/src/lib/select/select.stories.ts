import { SelectComponent, SelectModule } from './select.component';
import { Meta, moduleMetadata, Story } from '@storybook/angular';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { action } from '@storybook/addon-actions';

export default {
  title: 'atoms/forms/select',
  component: SelectComponent,
  decorators: [moduleMetadata({ imports: [SelectModule, BrowserAnimationsModule] })],
} as Meta<SelectComponent<unknown>>;

const template: Story<SelectComponent<string>> = () => ({
  props: {
    valueChange: action('Value change'),
  },
  template: `<gro-select placeholder="Select" (valueChange)="valueChange($event)">
    <gro-option>One</gro-option>
    <gro-option>Two</gro-option>
    <gro-option>Three</gro-option>
</gro-select>`,
});

export const Basic = template.bind({});
