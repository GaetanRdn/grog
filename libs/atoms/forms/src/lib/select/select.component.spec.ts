import { fakeAsync, TestBed, tick, waitForAsync } from '@angular/core/testing';
import { SelectComponent } from './select.component';
import { Component, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { TemplateLookup } from '@grorg/tests';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { EqualsFn, Nullable } from '@grorg/types';
import { OptionComponent } from '../option/option.component';
import { FormControl, ReactiveFormsModule } from '@angular/forms';

describe('Select', () => {
  beforeEach(
    waitForAsync(() => {
      TestBed.configureTestingModule({
        declarations: [
          SelectComponent,
          BasicSelectComponent,
          OptionComponent,
          ReactiveSelectComponent,
          ComplexSelectComponent,
        ],
        imports: [NoopAnimationsModule, ReactiveFormsModule],
        schemas: [CUSTOM_ELEMENTS_SCHEMA],
      }).compileComponents();
    })
  );

  describe('Basic', () => {
    let templateLookup: TemplateLookup<BasicSelectComponent>;

    beforeEach(() => (templateLookup = new TemplateLookup(BasicSelectComponent)));

    test('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create with placeholder', () => {
      templateLookup.hostComponent.placeholder = 'Select';
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should open options on click', () => {
      templateLookup.detectChanges();

      openPanel(templateLookup);

      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should close options on click out', () => {
      templateLookup.detectChanges();
      openPanel(templateLookup);
      expect(templateLookup.query('.gro-options-wrapper')).toBeTruthy();

      templateLookup.query('.click-out').click();
      templateLookup.detectChanges();

      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when select option then check', () => {
      // GIVEN
      templateLookup.detectChanges();
      openPanel(templateLookup);

      // WHEN
      templateLookup.query('gro-option').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.value).toEqual('One');
    });

    test('should create with initial value', fakeAsync(() => {
      // WHEN
      templateLookup.hostComponent.value = 'Two';
      templateLookup.detectChanges();
      tick(50);

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.getComponent(SelectComponent).options.toArray()[1].selected).toBe(true);
    }));

    test('should be disabled', () => {
      // WHEN
      templateLookup.hostComponent.disabled = true;
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should not opened options panel on click if disabled', () => {
      // GIVEN
      templateLookup.hostComponent.disabled = true;
      templateLookup.detectChanges();

      // WHEN
      openPanel(templateLookup);

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Reactive', () => {
    let templateLookup: TemplateLookup<ReactiveSelectComponent>;

    beforeEach(() => (templateLookup = new TemplateLookup(ReactiveSelectComponent)));

    test('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create with initial control value', fakeAsync(() => {
      templateLookup.hostComponent.control.setValue('Three');
      templateLookup.detectChanges();
      tick(50);

      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.getComponent(SelectComponent).viewValue).toEqual('Three');
      expect(templateLookup.getComponent(SelectComponent).options.toArray()[2].selected).toBe(true);
    }));

    test('should update control value on option selection', () => {
      templateLookup.detectChanges();

      openPanel(templateLookup);
      templateLookup.query('gro-option').click();
      templateLookup.detectChanges();

      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.control.value).toEqual('One');
    });

    test('should be disabled', () => {
      // WHEN
      templateLookup.hostComponent.control.disable();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Complex value', () => {
    let templateLookup: TemplateLookup<ComplexSelectComponent>;

    beforeEach(() => (templateLookup = new TemplateLookup(ComplexSelectComponent)));

    test('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create with initial value', () => {
      templateLookup.hostComponent.value = { value: 2, label: 'Two', toString: () => 'Two' };
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when select option then check', () => {
      // GIVEN
      templateLookup.detectChanges();
      openPanel(templateLookup);

      // WHEN
      templateLookup.query('gro-option').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(
        templateLookup.getComponent(SelectComponent).equals(templateLookup.hostComponent.value, {
          value: 1,
          label: 'One',
          toString: () => 'One',
        })
      ).toBe(true);
    });
  });

  function openPanel(templateLookup: TemplateLookup<unknown>): void {
    templateLookup.query('gro-select').click();
    templateLookup.detectChanges();
  }
});

@Component({
  template: ` <gro-select [placeholder]="placeholder" [(value)]="value" [disabled]="disabled">
      <gro-option>One</gro-option>
      <gro-option>Two</gro-option>
      <gro-option>Three</gro-option> </gro-select
    ><span class="click-out"></span>`,
})
class BasicSelectComponent {
  public placeholder = '';

  public value: Nullable<string> = null;

  public disabled = false;
}

@Component({
  template: ` <gro-select [(value)]="value" [equals]="equals">
      <gro-option *ngFor="let option of options" [value]="option">{{ option.label }}</gro-option> </gro-select
    ><span class="click-out"></span>`,
})
class ComplexSelectComponent {
  public readonly options = [
    { value: 1, label: 'One', toString: () => 'One' },
    { value: 2, label: 'Two', toString: () => 'Two' },
    { value: 3, label: 'Three', toString: () => 'Three' },
  ] as const;

  public value: Nullable<{ value: number; label: string; toString: () => string }> = null;

  public readonly equals: EqualsFn<{ value: number; label: string; toString: () => string }> = (val1, val2) =>
    val1?.value === val2?.value;
}

@Component({
  template: ` <gro-select [placeholder]="placeholder" [formControl]="control">
      <gro-option>One</gro-option>
      <gro-option>Two</gro-option>
      <gro-option>Three</gro-option> </gro-select
    ><span class="click-out"></span>`,
})
class ReactiveSelectComponent {
  public placeholder = '';

  public control: FormControl = new FormControl();
}
