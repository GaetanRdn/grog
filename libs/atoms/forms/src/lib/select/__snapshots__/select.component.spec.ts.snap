// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Basic should be disabled 1`] = `
<gro-select
  class="gro-disabled"
>
  <div
    class="gro-select-wrapper"
  >
    <span>
      
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Basic should close options on click out 1`] = `
<gro-select>
  <div
    class="gro-select-wrapper"
  >
    <span>
      
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  <div
    class="gro-options-wrapper ng-trigger ng-trigger-slideDownUp ng-animating"
    style="height:12rem;0:height;"
  >
    <gro-option>
      One
    </gro-option>
    <gro-option>
      Two
    </gro-option>
    <gro-option>
      Three
    </gro-option>
  </div>
  
</gro-select>
`;

exports[`Select Basic should create 1`] = `
<gro-select>
  <div
    class="gro-select-wrapper"
  >
    <span>
      
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Basic should create with initial value 1`] = `
<gro-select>
  <div
    class="gro-select-wrapper"
  >
    <span>
      Two
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Basic should create with placeholder 1`] = `
<gro-select>
  <div
    class="gro-select-wrapper"
  >
    <span>
      Select
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Basic should not opened options panel on click if disabled 1`] = `
<gro-select
  class="gro-disabled"
>
  <div
    class="gro-select-wrapper"
  >
    <span>
      
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Basic should open options on click 1`] = `
<gro-select
  class="gro-opened"
>
  <div
    class="gro-select-wrapper"
  >
    <span>
      
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  <div
    class="gro-options-wrapper ng-trigger ng-trigger-slideDownUp ng-animating"
    style="height:12rem;0:height;"
  >
    <gro-option>
      One
    </gro-option>
    <gro-option>
      Two
    </gro-option>
    <gro-option>
      Three
    </gro-option>
  </div>
  
</gro-select>
`;

exports[`Select Basic when select option then check 1`] = `
<gro-select>
  <div
    class="gro-select-wrapper"
  >
    <span>
      One
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  <div
    class="gro-options-wrapper ng-trigger ng-trigger-slideDownUp ng-animating"
    style="height:12rem;0:height;"
  >
    <gro-option
      class="gro-selected"
    >
      One
    </gro-option>
    <gro-option>
      Two
    </gro-option>
    <gro-option>
      Three
    </gro-option>
  </div>
  
</gro-select>
`;

exports[`Select Complex value should create 1`] = `
<gro-select>
  <div
    class="gro-select-wrapper"
  >
    <span>
      
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Complex value should create with initial value 1`] = `
<gro-select>
  <div
    class="gro-select-wrapper"
  >
    <span>
      Two
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Complex value when select option then check 1`] = `
<gro-select>
  <div
    class="gro-select-wrapper"
  >
    <span>
      One
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  <div
    class="gro-options-wrapper ng-trigger ng-trigger-slideDownUp ng-animating"
    style="height:12rem;0:height;"
  >
    <gro-option
      class="gro-selected"
      style=""
    >
      One
    </gro-option>
    <gro-option
      style=""
    >
      Two
    </gro-option>
    <gro-option
      style=""
    >
      Three
    </gro-option>
    
  </div>
  
</gro-select>
`;

exports[`Select Reactive should be disabled 1`] = `
<gro-select
  class="gro-disabled ng-untouched ng-pristine"
>
  <div
    class="gro-select-wrapper"
  >
    <span>
      
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Reactive should create 1`] = `
<gro-select
  class="ng-untouched ng-pristine ng-valid"
>
  <div
    class="gro-select-wrapper"
  >
    <span>
      
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Reactive should create with initial control value 1`] = `
<gro-select
  class="ng-untouched ng-pristine ng-valid"
>
  <div
    class="gro-select-wrapper"
  >
    <span>
      Three
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  
</gro-select>
`;

exports[`Select Reactive should update control value on option selection 1`] = `
<gro-select
  class="ng-valid ng-touched ng-dirty"
>
  <div
    class="gro-select-wrapper"
  >
    <span>
      One
    </span>
    <gro-material-icon>
      expand_more
    </gro-material-icon>
  </div>
  <div
    class="gro-options-wrapper ng-trigger ng-trigger-slideDownUp ng-animating"
    style="height:12rem;0:height;"
  >
    <gro-option
      class="gro-selected"
    >
      One
    </gro-option>
    <gro-option>
      Two
    </gro-option>
    <gro-option>
      Three
    </gro-option>
  </div>
  
</gro-select>
`;
